src/stream/src/executor/managed_state/top_n/top_n_state_new.rs

```diff
@@ -131,65 +124,22 @@ impl<S: StateStore> ManagedTopNStateNew<S> {
         TopNStateRow::new(pk_ordered, row)
     }
 
-    /// This function will return the rows at the position of `offset` and (`offset` + `limit` - 1),
-    /// which forms the range `[start_row, end_row]` of the top-N range .
-    /// When `offset` is 0 the `start_row` will be the first row in state table;
-    /// When `limit` is 0 the `end_row` will be an invalid row.
+    /// This function will return the rows in the range of [`offset`, `offset` + `limit`),
     pub async fn find_range(
         &self,
         offset: usize,
-        limit: usize,
+        num_limit: usize,
         epoch: u64,
-    ) -> StreamExecutorResult<(TopNStateRow, TopNStateRow)> {
+    ) -> StreamExecutorResult<Vec<TopNStateRow>> {
         let state_table_iter = self.state_table.iter(epoch).await?;
         pin_mut!(state_table_iter);
-
-        let invalid_row = TopNStateRow {
-            ordered_key: None,
-            row: Row::default(),
-        };
-        let mut first_row = invalid_row.clone();
-        if let Some(next_res) = state_table_iter.next().await {
-            first_row = self.get_topn_row(next_res);
-        }
-
-        // fetch start row
-        let mut start_row = first_row;
-        for i in 0..offset {
-            match state_table_iter.next().await {
-                Some(next_res) => {
-                    if i == offset - 1 {
-                        start_row = self.get_topn_row(next_res);
-                    }
-                }
-                None => {
-                    start_row = invalid_row.clone();
-                    break;
-                }
-            }
-        }
-
-        // fetch end row
-        if limit == 0 {
-            return Ok((start_row, invalid_row.clone()));
-        }
-        let mut end_row = start_row.clone();
-        let num_limit = limit - 1;
-        for i in 0..num_limit {
-            match state_table_iter.next().await {
-                Some(next_res) => {
-                    if i == num_limit - 1 {
-                        end_row = self.get_topn_row(next_res);
-                    }
-                }
-                None => {
-                    end_row = invalid_row.clone();
-                    break;
-                }
-            }
-        }
-
-        Ok((start_row, end_row))
+        let rows = state_table_iter
+            .skip(offset)
+            .take(num_limit)
+            .map(|x| self.get_topn_row(x))
```

Comment:
We should not directly unwrap this storage result.

Reply:
Do we really need owned `Row` by `Cow::into_owned` here? Seems generating outputs only require datum refs.